internal/tfr/getter.go

```diff
@@ -0,0 +1,259 @@
+)
+
+// TerraformRegistryGetter is a Getter (from go-getter) implementation that will download from the terraform module
+// registry. This supports getter URLs encoded in the following manner:
+//
+// tfr://REGISTRY_DOMAIN/MODULE_PATH?version=VERSION
+//
+// Where the REGISTRY_DOMAIN is the terraform registry endpoint (e.g., registry.terraform.io), MODULE_PATH is the
+// registry path for the module (e.g., terraform-aws-modules/vpc/aws), and VERSION is the specific version of the module
+// to download (e.g., 2.2.0).
+//
+// This protocol will use the Module Registry Protocol (documented at
+// https://www.terraform.io/docs/internals/module-registry-protocol.html) to lookup the module source URL and download
+// it.
+//
+// Authentication to private module registries is handled via environment variables. The authorization API token is
+// expected to be provided to Terragrunt via the TG_TF_REGISTRY_TOKEN environment variable. This token can be any
+// registry API token generated on Terraform Cloud / Enterprise.
+//
+// MAINTAINER'S NOTE: Ideally we implement the full credential system that terraform uses as part of `terraform login`,
+// but all the relevant packages are internal to the terraform repository, thus making it difficult to use as a
+// library. For now, we keep things simple by supporting providing tokens via env vars and in the future, we can
+// consider implementing functionality to load credentials from terraform.
+type TerraformRegistryGetter struct {
+	client *getter.Client
+}
+
+// SetClient allows the getter to know what getter client (different from the underlying HTTP client) to use for
+// progress tracking.
+func (tfrGetter *TerraformRegistryGetter) SetClient(client *getter.Client) {
+	tfrGetter.client = client
+}
+
+// Context returns the go context to use for the underlying fetch routines. This depends on what client is set.
+func (tfrGetter *TerraformRegistryGetter) Context() context.Context {
+	if tfrGetter == nil || tfrGetter.client == nil {
+		return context.Background()
+	}
+	return tfrGetter.client.Ctx
+}
+
+// ClientMode returns the download mode based on the given URL. Since this getter is designed around the Terraform
+// module registry, we always use Dir mode so that we can download the full Terraform module.
+func (tfrGetter *TerraformRegistryGetter) ClientMode(u *url.URL) (getter.ClientMode, error) {
+	return getter.ClientModeDir, nil
+}
+
+// Get is the main routine to fetch the module contents specified at the given URL and download it to the dstPath.
+// This routine assumes that the srcURL points to the Terraform registry URL, with the Path configured to the module
+// path encoded as `:namespace/:name/:system` as expected by the Terraform registry. Note that the URL query parameter
+// must have the `version` key to specify what version to download.
+func (tfrGetter *TerraformRegistryGetter) Get(dstPath string, srcURL *url.URL) error {
+	ctx := tfrGetter.Context()
+
+	registryDomain := srcURL.Host
```

Comment:
NIT: worth defaulting the Host to `app.terraform.io` if not explicitly specified in `source`? Would allow for the pretty short URLs like `hashicorp/consul/aws`.

Reply:
Ah that's a good idea. Done: 0ad5c1e

Reply:
Nice! We should update the docs accordingly!

Reply:
Specifically, these docs: https://github.com/gruntwork-io/terragrunt/pull/1767/files#diff-eac47411aff1ec957598f585714c7199a1c1c31a710ef472868bae5abef38cf9R43.

Reply:
Done: d3531d9